### PR TITLE
Render improvements

### PR DESCRIPTION
**Pull request by sidorares**
WIP

added `Render.AddTraps` and `Render.SetPictureTransform` + render enums

the changes here mainly because I'm trying to get shape drawing code working. Initial attempt using Render.Triangles + triangulation with https://github.com/mikolalysenko/triangulate-polyline didn't work very well because Render not stitching triangle edges properly ( or maybe I'm missing some flags ):

screen shot 2015-11-04 at 4 54 37 pm

Next attempt - split shape into horisontal trapezoids using https://github.com/jahting/pnltri.js
```js
  var myTrapezoider = new PNLTRI.Trapezoider( new PNLTRI.PolygonData(example_data) );
  myTrapezoider.trapezoide_polygon();
  var traps = myTrapezoider.queryStructure.trapArray;
```
+ and code similar to https://github.com/jahting/pnltri.js/blob/master/test/TestHelpers.js#L155-L189

### CODE REVIEW

File: lib/ext/render.js [hunk 1]
```diff
@@ -152,34 +152,69 @@ exports.requireExt = function(display, callback)
...
-                return parseInt(f*65536);
+                return parseInt(f*65535);
             }
 
+            ext.SetPictureTransform = function(pid, matrix) {
+              X.seq_num++;
+              var format = 'CCSL';
+              if (typeof matrix.length !== 9)
+                throw 'Render.SetPictureTransform: incorrect transform matrix. Must be array of 9 numbers';
+              var params = [ext.majorOpcode, 28, 11, pid];
```
Comment (santigimeno) [inline lib/ext/render.js]: Shouldn't we just throw ASAP?

File: lib/ext/render.js [hunk 2]
```diff
@@ -152,34 +152,69 @@ exports.requireExt = function(display, callback)
...
+              for (var i=0; i < 9; ++i) {
+                format += 'L';
+                if (typeof matrix[i] !== 'number')
+                  throw 'Render.SetPictureTransform: matrix element must be a number';
+                params.push(floatToFix(matrix[i]));
+              }
+              X.pack_stream.pack(format, params);
```
Comment (santigimeno) [inline lib/ext/render.js]: Maybe increment seq_num only when we now we're sending the request

File: lib/ext/render.js [hunk 3]
```diff
@@ -152,34 +152,69 @@ exports.requireExt = function(display, callback)
...
-                var params = [ext.majorOpcode, 30, reqLen, pid];
-                /*
-                if (name == 'convolution')
-                {
-                    reqLen += 2;
-                    format += 'L';
-                    params.push(floatToFix(filterParams));
+                var reqLen = 2;
+                var format = 'CCSLSxxp';
+                var params = [ext.majorOpcode, 30, reqLen, pid, name.length, name + '\0'];
```
Comment (santigimeno) [inline lib/ext/render.js]: same comment here about seq_num

**Comment by santigimeno** [comment]
Code looks fine. I'll try to check the protocol implementation later

File: lib/ext/render.js [hunk 4]
```diff
@@ -152,34 +152,69 @@ exports.requireExt = function(display, callback)
...
 
+            ext.SetPictureTransform = function(pid, matrix) {
+              X.seq_num++;
+              var format = 'CCSL';
+              if (typeof matrix.length !== 9)
+                throw 'Render.SetPictureTransform: incorrect transform matrix. Must be array of 9 numbers';
+              var params = [ext.majorOpcode, 28, 11, pid];
+              for (var i=0; i < 9; ++i) {
+                format += 'L';
+                if (typeof matrix[i] !== 'number')
```
Comment (santigimeno) [inline lib/ext/render.js]: The format can be defined in one step as it HAS to have 9 'L' anyways

File: lib/ext/render.js [hunk 5]
```diff
@@ -152,34 +152,69 @@ exports.requireExt = function(display, callback)
...
+                } else if (name == 'convolution') {
+                   if (filterParams.length < 2 || filterParams[0]*filterParams[1] !== filterParams.length + 2) {
+                     throw 'Render.SetPictureFilter: "convolution" - incorrect matrix dimensions. Must be flat array [ w, h, elem1, elem2, ... ]';
+                   }
+                   for (var i=0; i < filterParams.length; ++i) {
+                      format += 'L';
+                      params.push(floatToFix(filterParams[i]));
+                   }
+                } else if (name == 'binomial' || name == 'gaussian') {
+                   if (filterParams.length < 2) {
```
Comment (santigimeno) [inline lib/ext/render.js]: s/==/===

Comment (santigimeno) [inline lib/ext/render.js]: if (filterParams.length !== 1)

File: lib/ext/render.js [hunk 6]
```diff
@@ -565,5 +614,112 @@ exports.requireExt = function(display, callback)
...
+            Normal: 1,
+            Pad: 2,
+            Reflect: 3
+          };
+
+          ext.Subpixel = {
+            Unknown:       0,
+            HorizontalRGB: 1,
+            HorizontalBGR: 2,
+            VerticalRGB	 : 3,
```
Comment (santigimeno) [inline lib/ext/render.js]: format

**Comment by sidorares** [comment]
Yes, noticed some of those myself.
I'll probably add a test that just requires all extensions so at least CI checks it's valid js

Good point about incrementing seq_num after all parameters processed.

Also I noticed color values are used very inconsistently across requests - sometimes thay are float values, sometimes fixed point ( 16 bit int), sometimes array of [r, g, b, a] sometimes not. The change is going to be backwards incompatible, part of 2.x release

**Comment by sidorares** [comment]
merging this, more to follow

**Pull request merged**